File: CoreAutomata/Region.cs
using System;

namespace CoreAutomata
{
    public class Region
    {
        public int X { get; set; }
        public int Y { get; set; }
        public int W { get; set; }
        public int H { get; set; }

        public int R => X + W;

        public int B => Y + H;

        public Location Location => new Location(X, Y);

        public Region(int X, int Y, int W, int H)
        {
            this.X = X;
            this.Y = Y;
            this.W = W;
            this.H = H;
        }

        /// <summary>
        /// Restrict another Region to be within the bounds of this Region.
        /// </summary>
        public Region Clip(Region Region)
        {
            var left = Region.X.Clip(X, R - 1);
            var right = Region.R.Clip(X + 1, R);
            var top = Region.Y.Clip(Y, B - 1);
            var bottom = Region.B.Clip(Y + 1, B);

            return new Region(left, top, right - left, bottom - top);
        }

        public bool Contains(Region Region)
        {
            return X <= Region.X
                   && Y <= Region.Y
                   && R >= Region.R
                   && B >= Region.B;
        }

        public void WaitVanish(IPattern Image, int? Timeout = null) => AutomataApi.WaitVanish(this, Image, Timeout);

        public bool Exists(IPattern Image, int? Timeout = null, double? Similarity = null) => AutomataApi.Exists(this, Image, Timeout, Similarity);

        public IPattern GetPattern() => AutomataApi.GetCopy(this);

        public void Click()
        {
            var center = new Location(X + W / 2, Y + H / 2);

            center.Click();
        }

        public static Region operator *(Region Region, double Scale)
        {
            return new Region(
                (Region.X * Scale).Round(),
                (Region.Y * Scale).Round(),
                (Region.W * Scale).Round(),
                (Region.H * Scale).Round());
        }

        public void Highlight(TimeSpan Duration)
        {
            AutomataApi.Highlight(this, Duration);
        }

        public override string ToString() => $"({X}, {Y}) {W}x{H}";
    }
}

File: app/src/main/java/com/mathewsachin/fategrandautomata/core/Region.kt
package com.mathewsachin.fategrandautomata.core

import kotlin.math.roundToInt

data class Region(val X: Int, val Y: Int, val Width: Int, val Height: Int) {
    init {
        require(Width > 0) { "Width must be positive" }
        require(Height > 0) { "Height must be positive" }
    }

    constructor(Location: Location, Size: Size) : this(Location.X, Location.Y, Size.Width, Size.Height)

    operator fun times(scale: Double): Region {
        return Region(
            (X * scale).roundToInt(),
            (Y * scale).roundToInt(),
            (Width * scale).roundToInt(),
            (Height * scale).roundToInt()
        )
    }

    val location get() = Location(X, Y)

    val size get() = Size(Width, Height)

    val center get() = Location(X + Width / 2, Y + Height / 2)

    val right get() = X + Width

    val bottom get() = Y + Height

    fun clip(Region: Region): Region {
        val left = Region.X.coerceIn(X, right - 1)
        val right = Region.right.coerceIn(X + 1, right)
        val top = Region.Y.coerceIn(Y, bottom - 1)
        val bottom = Region.bottom.coerceIn(Y + 1, bottom)

        return Region(left, top, right - left, bottom - top)
    }

    fun contains(Region: Region): Boolean {
        return X <= Region.X
                && Y <= Region.Y
                && right >= Region.right
                && bottom >= Region.bottom
    }
}